Source/Python/adapter/persistence/in_memory_repositories.py

```diff
@@ -86,18 +87,42 @@ def delete(self, traveler_id: PrefixedUUID) -> None:
 
 class InMemoryEventRepository(EventRepository):
     _inner_repo: _InMemoryIdentifiedEntityRepository
+    _event_ids_by_location_id: Dict[PrefixedUUID, Set[PrefixedUUID]]
+    _event_ids_by_traveler_id: Dict[PrefixedUUID, Set[PrefixedUUID]]
 
     def __init__(self) -> None:
         self._inner_repo = _InMemoryIdentifiedEntityRepository(Event)
+        self._event_ids_by_location_id = defaultdict(set)
+        self._event_ids_by_traveler_id = defaultdict(set)
 
     def save(self, event: Event) -> None:
         self._inner_repo.save(event)
+        for location_id in event.affected_locations:
+            self._event_ids_by_location_id[location_id].add(event.id)
+        for traveler_id in event.affected_travelers:
+            self._event_ids_by_traveler_id[traveler_id].add(event.id)
 
     def retrieve(self, event_id: PrefixedUUID) -> Event:
         return self._inner_repo.retrieve(event_id)
 
-    def retrieve_all(self) -> Set[Event]:
-        return self._inner_repo.retrieve_all()
+    def retrieve_all(self, *, location_id: PrefixedUUID = None, traveler_id: PrefixedUUID = None) -> Set[Event]:
+        if location_id is None and traveler_id is None:
+            # Neither filter provided, return all
+            return self._inner_repo.retrieve_all()
+
+        events_linked_to_provided_location_id = self._event_ids_by_location_id[location_id] if location_id is not None else set()
+        events_linked_to_provided_traveler_id = self._event_ids_by_traveler_id[traveler_id] if traveler_id is not None else set()
```

Comment:
This could be replaced with a `self._event_ids_by_traveler_id.get(traveler_id, set())`, as None will just result in using the default. Similar can be done for the travelers.

Reply:
Fixed